src/pymor/algorithms/phdmd.py

```diff
@@ -0,0 +1,564 @@
+                    NumpyMatrixOperator(initial_J[space_dim:, space_dim:])
+                ]
+            ])
+
+        assert initial_J.source == block_space and initial_J.range == block_space
+        initial_R, spsd_data = _initial_psd_symmetric_procrustes(T, initial_J.apply(T) - Z)
+        data['spsd_init_data'] = spsd_data
+    elif initial_J is None and initial_R is None:
+        logger.info('Solving weighted problem to get initial operators...')
+        initial_J, initial_R, weighted_data = _weighted_phdmd(T, Z, rtol=weighted_rtol)
+        data['weighted_init_data'] = weighted_data
+
+    assert isinstance(initial_J, BlockOperator) and isinstance(initial_R, BlockOperator)
+    assert initial_J.source == initial_R.source == initial_J.range == initial_R.range
+
+    J = initial_J
+    R = initial_R
+
+    logger.info('Running pH DMD algorithm...')
+
+    sing_vals = eigh(T.inner(T), eigvals_only=True)
+    L = max(sing_vals)
+    mu = min(sing_vals)
+    sing_ratio = mu / L
+
+    alphas = [initial_alpha]
+    betas = []
+    abs_errs = [np.linalg.norm((Z - (J - R).apply(T)).to_numpy(), 'fro')]
+    rel_errs = [abs_errs[0] / np.linalg.norm(Z.to_numpy(), 'fro')]
+    update_norms = []
+    procrustes_data = []
+
+    Q = R
+    iteration = 0
+    update_norm = np.inf
+    while True:
+        logger.info(f'Starting pH DMD algorithm iteration {iteration} with '
+                    f'current errors abs: {abs_errs[-1]} rel: {rel_errs[-1]}...')
+
+        if iteration >= miniter:
+            if abs_errs[-1] < atol:
+                logger.info(f'pH DMD converged in {iteration} iterations because '
+                            f'absolute error tolerance of {atol} was reached.')
+                break
+            if rel_errs[-1] < rtol:
+                logger.info(f'pH DMD converged in {iteration} iterations because '
+                            f'relative error tolerance of {rtol} was reached.')
+                break
+            if update_norm < rtol:
+                logger.info(f'pH DMD converged in {iteration} iterations because '
+                            f'relative update tolerance of {rtol} was reached.')
+                break
+            if iteration >= maxiter:
+                logger.info(f'Maximum iterations reached. Failed to converge after {iteration} iterations.')
+                raise PHDMDError('Failed to converge after the maximum amount of iterations.')
+
+        iteration += 1
+
+        R_prev = R
+        J_prev = J
+        last_alpha = alphas[-1]
+
+        Z_1 = Z + R.apply(T)
+        logger.info('Solving skew-symmetric Procrustes subproblem...')
+        J, skew_procrustes_data = _skew_symmetric_procrustes(T, Z_1, rtol=skew_procrustes_rtol)
+        procrustes_data.append(skew_procrustes_data)
+        Z_2 = J.apply(T) - Z
+        G = Q.apply(T).to_numpy().T @ T.to_numpy() - Z_2.to_numpy().T @ T.to_numpy()
+
+        R = _project_spsd(to_matrix(Q) - (1. / L) * G, dim=space_dim, id=space_id)
```

Comment:
`to_numpy` and `to_matrix` should be avoided if the goal is to have an algorithm for high-dimensional data.

Reply:
As far as I understand, the pH DMD algorithm is never applied to high-dimensional state data. The general strategy is to reduce the states before passing them into pH DMD, hence the `X`/`Xdot` matrices always fulfill `num_time_snapshots >> state_dim` and similarly `num_time_snapshots >> num_in-/outputs` for `U` and `Y`.

Reply:
Ok, if that is true, then I would either:

1. add a reduction step that transforms `VectorArrays` `X` and `Xdot` to low-dimensional NumPy arrays, or
2. only accept NumPy arrays to the algorithm.

We actually already have an algorithm that only uses NumPy arrays (in `bernoulli.py`), so the second option would not be so out of line.